### DISCUSSION TITLE
XInput Controller

### DISCUSSION THREAD

**Discussion by 1dawideq1** [amerkoleci/Vortice.Windows#208] [Q&A]
Hello,

my goal is to create an virtual controller, whose values are set by joysticks, that are connected via Ethernet. Is there any way to initialize such Controller?

Best regards

**Comment by amerkoleci** [comment]
Not sure how this is related to XInput and/or Vortice.Windows in general?

**Reply by 1dawideq1** [reply]
The joysticks have to be detected as Xbox Controller. That is where it gets complicated, because joysticks are connected to a controls board, which is plugged into PC via ethernet. Now I need somehow to create an "abstract" controller, which will be detected by Windows as XboxController. And it will pass values from the joysticks

**Reply by amerkoleci** [reply]
Vortice just wraps XInput native API, if it is not supported by native API then its not supported by Vortice as well

**Reply by 1dawideq1** [reply]
Ah ok. Do you have any idea how I could achieve this?

**Reply by amerkoleci** [reply]
Actually no, sorry, never had the need to do something like this :)

**Reply by h1cks** [reply]
> 

The way I have done something here is to create an abstract controller layer, the layer executes movement on the player as Actions.  This layer connects to the player avatar.

The abstract layer is referenced via a mapping layer, this mapping layer takes a device input and its output and maps it to an Action.

What you should do here is take your ethernet input and map that to a mapping layer into the abstract layer.  You dont need to then fake a hardware device such as a XBox Controller.

Example:

Control board -> Ethernet -> Ethernet receiving library -> output to Mapping Layer -> Abstract Controller -> Player.
Xbox Controller -> output to Mapping Layer -> Abstract Controller -> Player.
Keyboard -> output to Mapping Layer -> Abstract Controller -> Player.
Joystick -> output to Mapping Layer -> Abstract Controller -> Player.

IN my code, I enumerate all devices,  and they can be assigned to an Abstract Controller.  The abstraction layer is crucial in normalising various different gaming inputs.

**Comment by Axiverse** [comment]
Either use in C++ or port https://github.com/shauleiz/vXboxInterface/ to C#.

See https://stackoverflow.com/questions/11711947/how-do-i-create-a-virtual-gamepad

What you want to do is in driver land and isn't really a directx issue.
